### DISCUSSION TITLE
maximum connections allowed

### DISCUSSION THREAD

**Discussion by akcopy** [milvus-io/milvus#30242] [Q&A and General discussion] [answered]
I am trying to integrate similarity search into my service at the moment and was wondering if there's a hard limit on the number of connections that can be made into a single Milvus instance?  Also, are simultaneous connections with multiple threads possible with using the Langchain library for querying?

**Comment by xiaofan-luan** [comment]
For 2.3.x, our recommendation is not beyond 10000 collections.

For multi tenancy support, best way is to use partitionkey features for logical isolation

**Comment by akcopy** [comment]
sorry about the confusion, but i was talking more about multiple connections to the same database/partitions. For e.g. if I had 1000 threads all calling the following using Langchain, would it still work? I only tested it with 1 connection, was just wondering if there's a limit when I try to scale this up.

vector_store = Milvus(
    embedding_function=embeddings,
    connection_args=connection_args,
    collection_name=COLLECTION_NAME)
    
docs = vector_store.similarity_search(query)

**Reply by xiaofan-luan** [reply]
usually, this should require a filter condition to specify partitonkey == xx

**Reply by akcopy** [reply]
the documents would all be preloaded into the database. So I won't be adding any new documents to the database, merely doing similarity searches (with the connections). I'm not having any partitions in the database. Would it be possible for a single Milvus instance to support thousands of simultaneous similarity searches conducted using different threads?

**Comment by xiaofan-luan** [comment]
https://python.langchain.com/docs/integrations/vectorstores/milvus#per-user-retrieval 

the new langchain docuemnt should updated

**Reply by akcopy** [reply]
thanks - but I was more looking for simultaneous Langchain connections to the same dataset/collection (i.e. all the 1000+ users are searching in the same fixed collection in a single milvus instance) - is that something that has been tested?

**Comment by yhmo** [comment, marked as answer]
- Milvus server is based on a grpc server, the server doesn't limit the number of connections. But, the milvus proxy node has a task queue, the task queue has limit=1024. So, if you have lots of clients with simultaneous connections to send requests, and the requests are not processed in time, the proxy will reject requests if the queue is full.

- In the Langchian source code, the logic for client connections: https://github.com/langchain-ai/langchain/blob/9e95699277fe0db3bfee1654276a43bfba9ecc64/libs/community/langchain_community/vectorstores/milvus.py#L241

We can see the milvus client is implemented as a class:
`class Milvus(VectorStore)`
In the _create_connection_alias method, we can see these lines:
```
from pymilvus import MilvusException, connections
for con in connections.list_connections():
  addr = connections.get_connection_addr(con[0])
  if (addr["address"] == given_address):
    return con[0]
```

The "connections" is a global object in the current process. It manages all the connections for the current process.
So, if you want to create "simultaneous connections", you should use "multi processes". In each process declare a Milvus vectorstore object. All these Milvus vectorstore are connecting to the same MIlvus server, they are in different processes so they own different connections.

**Reply by akcopy** [reply]
Does it mean I can bypass the limit by having more proxy nodes?

**Reply by yhmo** [reply]
If the proxy task queue is full, that means the bottleneck is in the background components(querynode/datanode). So, no matter how many proxies are added, their task queues will eventually be full.
